src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.DebugVerifier.cs

```diff
@@ -0,0 +1,208 @@
+            }
+
+            public static void Verify(NullableWalker walker, BoundNode node)
+            {
+                var verifier = new DebugVerifier(walker._analyzedNullabilityMapOpt);
+                verifier.Visit(node);
+                // Can't just remove nodes from _topLevelNullabilityMap because nodes can be reused.
+                Debug.Assert(verifier._analyzedNullabilityMap.Count == verifier._visitedNodes.Count, $"Visited {verifier._visitedNodes.Count} nodes, expected to visit {verifier._analyzedNullabilityMap.Count}");
+                verifier.Free();
+            }
+
+            private void Free()
+            {
+                _analyzedNullabilityMap.Free();
+            }
+
+            protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpression node)
+            {
+                VerifyNode(node);
+                return (BoundExpression)base.Visit(node);
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                if (node is BoundExpression expr)
+                {
+                    return VisitExpressionWithStackGuard(ref _recursionDepth, expr);
+                }
+                return base.Visit(node);
+            }
+
+            public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstructionAssignmentOperator node)
+            {
+                // PROTOTYPE(nullable-api): handle
+                return null;
+            }
+
+            public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationExpression node)
+            {
+                // If the delegate creation was resolved to a single static method, we do not examine the
+                // receiver of the method group as it's completely unused.
+                if (node.Argument is BoundMethodGroup group)
+                {
+                    VerifyNode(group);
+                    if (node.MethodOpt?.IsStatic == false)
+                    {
+                        Visit(group.ReceiverOpt);
+                    }
+
+                    return null;
+                }
+                else
+                {
+                    return base.VisitDelegateCreationExpression(node);
+                }
+            }
+
+            public override BoundNode VisitNewT(BoundNewT node)
+            {
+                // https://github.com/dotnet/roslyn/issues/33387 We're not currently
+                // examining child nodes correctly
+                if (node.InitializerExpressionOpt != null)
+                {
+                    VerifyNode(node.InitializerExpressionOpt, overrideSkippedExpression: true);
+                }
+
+                return null;
+            }
+
+            public override BoundNode VisitBadExpression(BoundBadExpression node)
+            {
+                // Regardless of what the BadExpression is, we need to add all of its direct children to the visited map. They
```

Comment:
>we need to add all of its direct children to the visited map [](start = 60, length = 60)

This seems unnecessary since it means visiting nodes in `NullableWalker` that we could otherwise skip. The consumer of the map should gracefully handle missing entries.

Reply:
Gracefully handling such entries isn't a good strategy. The whole purpose of this verifier is to ensure that aren't any missing entries, because missing entries means that information isn't being correctly recorded.

---
In reply to: [265303709](https://github.com/dotnet/roslyn/pull/34017#discussion_r265303709) [](ancestors = 265303709)

Reply:
I mean, yes consumers shouldn't crash when there are missing nodes. But as mentioned last time, this particular case is specifically handling entries that are or normally not in the map, mainly object initializers under bound bad expressions.

---
In reply to: [265372591](https://github.com/dotnet/roslyn/pull/34017#discussion_r265372591) [](ancestors = 265372591,265303709)

Reply:
> The consumer of the map should gracefully handle missing entries.

Could you clarify what that would mean? 

I would think there is useful/good nullable information for some expressions nested inside bad expressions (NullableWalker.VisitBadExpression does visit the child nodes). I'd think callers can only get that information from the map (they can't make it up).

---
In reply to: [265373564](https://github.com/dotnet/roslyn/pull/34017#discussion_r265373564) [](ancestors = 265373564,265372591,265303709)

Reply:
My original comment was incorrect. We should include the child expressions here.

---
In reply to: [265381256](https://github.com/dotnet/roslyn/pull/34017#discussion_r265381256) [](ancestors = 265381256,265373564,265372591,265303709)